container/lxd/lxd.go

```diff
@@ -176,67 +360,118 @@ func (manager *containerManager) CreateContainer(
 		logger.Infof("instance %q configured with %v network devices", name, nics)
 	}
 
-	spec := lxdclient.InstanceSpec{
-		Name:     name,
-		Image:    imageName,
-		Metadata: metadata,
-		Devices:  nics,
-		Profiles: profiles,
+	spec := api.ContainersPost{
+		Name: name,
+		ContainerPut: api.ContainerPut{
+			Profiles: profiles,
+			Devices:  nics,
+			Config:   metadata,
+		},
 	}
 
-	logger.Infof("starting instance %q (image %q)...", spec.Name, spec.Image)
-	callback(status.Provisioning, "Starting container", nil)
-	_, err = manager.client.AddInstance(spec)
+	logger.Infof("starting instance %q (image %q)...", spec.Name, imageName)
+	callback(status.Provisioning, "Creating container", nil)
+	op, err := manager.server.CreateContainerFromImage(imageServer, *image, spec)
 	if err != nil {
-		return
+		logger.Errorf("CreateContainer failed with %s", err)
+		return nil, nil, err
+	}
+
+	progress := func(op api.Operation) {
+		if op.Metadata == nil {
+			return
+		}
+		for _, key := range []string{"fs_progress", "download_progress"} {
+			value, ok := op.Metadata[key]
+			if ok {
+				callback(status.Provisioning, fmt.Sprintf("Retrieving image: %s", value.(string)), nil)
+				return
+			}
+		}
+	}
+	_, err = op.AddHandler(progress)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	op.Wait()
+	opInfo, err := op.GetTarget()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	if opInfo.StatusCode != api.Success {
+		return nil, nil, fmt.Errorf("LXD error: %s", opInfo.Err)
+	}
+
+	callback(status.Provisioning, "Container created, starting", nil)
+	err = manager.startInstance(name)
+	if err != nil {
+		if err := manager.removeInstance; err != nil {
+			logger.Errorf("Cannot remove failed instance: %s", err)
+		}
+		return nil, nil, err
 	}
 
 	callback(status.Running, "Container started", nil)
-	inst = &lxdInstance{name, manager.client}
-	return
+	inst = &lxdInstance{name, manager.server}
+	return inst, nil, nil
 }
 
 func (manager *containerManager) DestroyContainer(id instance.Id) error {
-	if manager.client == nil {
+	if manager.server == nil {
 		var err error
-		manager.client, err = ConnectLocal()
+		manager.server, err = ConnectLocal()
 		if err != nil {
 			return err
 		}
 	}
-	return errors.Trace(manager.client.RemoveInstances(manager.namespace.Prefix(), string(id)))
+	instances, err := manager.ListContainers()
+	if err != nil {
+		return err
+	}
+	for _, inst := range instances {
```

Comment:
shouldn't this loop be just
_, err := manager.server.DeleteContainer(string(id))

?

(having a side effect of "juju remove-machine 5" go and destroy all containers seems like a bad side effect. Let's make sure we have tests that DeleteContainer(X) doesn't touch any other containers)